src/Map/MapRasterizer_P.cpp

```diff
@@ -434,43 +437,32 @@ void OsmAnd::MapRasterizer_P::rasterizePolygon(
         canvas.drawPath(path, paint);
 }
 
-void OsmAnd::MapRasterizer_P::rasterizePolyline(
+bool OsmAnd::MapRasterizer_P::calcPathByTrajectory(
     const Context& context,
     SkCanvas& canvas,
-    const std::shared_ptr<const MapPrimitiviser::Primitive>& primitive,
-    bool drawOnlyShadow)
+    const QVector<PointI>& points31,
+    SkPath& path,
+    float offset = 0.0f) const
 {
-    const auto& points31 = primitive->sourceObject->points31;
-    const auto& area31 = context.area31;
-    const auto& env = context.env;
-
-    assert(points31.size() >= 2);
-
-    SkPaint paint = _defaultPaint;
-    if (!updatePaint(context, paint, primitive->evaluationResult, PaintValuesSet::Layer_1, false))
-        return;
-
-    bool ok;
+    int8_t dir = offset < 0 ? -1: 1;
+    offset = abs(offset);
+    bool shift = offset > std::numeric_limits<float>::epsilon() ? true : false;
 
-    ColorARGB shadowColor;
-    ok = primitive->evaluationResult.getIntegerValue(env->styleBuiltinValueDefs->id_OUTPUT_SHADOW_COLOR, shadowColor.argb);
-    if (!ok || shadowColor == ColorARGB::fromSkColor(SK_ColorTRANSPARENT))
-        shadowColor = context.shadowColor;
-
-    float shadowRadius;
-    ok = primitive->evaluationResult.getFloatValue(env->styleBuiltinValueDefs->id_OUTPUT_SHADOW_RADIUS, shadowRadius);
-    if (drawOnlyShadow && (!ok || shadowRadius <= 0.0f))
-        return;
-
-    SkPath path;
     int pointIdx = 0;
     bool intersect = false;
     int prevCross = 0;
     PointF vertex;
+    const auto& area31 = context.area31;
     const auto pointsCount = points31.size();
     auto pPoint = points31.constData();
     PointF pVertex;
     PointF tempVertex;
+    PointF correctedVertex;
+    const static uint8_t kLastPointCnt = 3;
+    // Could be implemented/extended custom simple deque to store only last 3 point for the originalPoints
+    std::deque<PointF> originalPoints;
```

Comment:
usually vector is clear & faster, I don't see what specific is used here for deque

Reply:
I would agree that you should use vector and preallocate it to max capacity